### PR TITLE
Sync dir containing CURRENT after RenameFile on CURRENT as much as possible

### PR DESCRIPTION
**Pull request by hx235**
**Context:** 
Below crash test revealed a bug that directory containing CURRENT file (short for `dir_contains_current_file` below) was not always get synced after a new CURRENT is created and being called with `RenameFile` as part of the creation. 

This bug exposes a risk that such un-synced directory containing the updated CURRENT can’t survive a host crash (e.g, power loss) hence get corrupted. This then will be followed by a recovery from a corrupted CURRENT that we don't want.  

The root-cause is that a nullptr `FSDirectory* dir_contains_current_file` sometimes gets passed-down to `SetCurrentFile()` hence in those case `dir_contains_current_file->FSDirectory::FsyncWithDirOptions()` will be skipped  (which otherwise will internally call`Env/FS::SyncDic()` )
```
./db_stress --acquire_snapshot_one_in=10000 --adaptive_readahead=1 --allow_data_in_errors=True --avoid_unnecessary_blocking_io=0 --backup_max_size=104857600 --backup_one_in=100000 --batch_protection_bytes_per_key=8 --block_size=16384 --bloom_bits=134.8015470676662 --bottommost_compression_type=disable --cache_size=8388608 --checkpoint_one_in=1000000 --checksum_type=kCRC32c --clear_column_family_one_in=0 --compact_files_one_in=1000000 --compact_range_one_in=1000000 --compaction_pri=2 --compaction_ttl=100 --compression_max_dict_buffer_bytes=511 --compression_max_dict_bytes=16384 --compression_type=zstd --compression_use_zstd_dict_trainer=1 --compression_zstd_max_train_bytes=65536 --continuous_verification_interval=0 --data_block_index_type=0 --db=$db --db_write_buffer_size=1048576 --delpercent=5 --delrangepercent=0 --destroy_db_initially=0 --disable_wal=0 --enable_compaction_filter=0 --enable_pipelined_write=1 --expected_values_dir=$exp --fail_if_options_file_error=1 --file_checksum_impl=none --flush_one_in=1000000 --get_current_wal_file_one_in=0 --get_live_files_one_in=1000000 --get_property_one_in=1000000 --get_sorted_wal_files_one_in=0 --index_block_restart_interval=4 --ingest_external_file_one_in=0 --iterpercent=10 --key_len_percent_dist=1,30,69 --level_compaction_dynamic_level_bytes=True --mark_for_compaction_one_file_in=10 --max_background_compactions=20 --max_bytes_for_level_base=10485760 --max_key=10000 --max_key_len=3 --max_manifest_file_size=16384 --max_write_batch_group_size_bytes=64 --max_write_buffer_number=3 --max_write_buffer_size_to_maintain=0 --memtable_prefix_bloom_size_ratio=0.001 --memtable_protection_bytes_per_key=1 --memtable_whole_key_filtering=1 --mmap_read=1 --nooverwritepercent=1 --open_metadata_write_fault_one_in=0 --open_read_fault_one_in=0 --open_write_fault_one_in=0 --ops_per_thread=100000000 --optimize_filters_for_memory=1 --paranoid_file_checks=1 --partition_pinning=2 --pause_background_one_in=1000000 --periodic_compaction_seconds=0 --prefix_size=5 --prefixpercent=5 --prepopulate_block_cache=1 --progress_reports=0 --read_fault_one_in=1000 --readpercent=45 --recycle_log_file_num=0 --reopen=0 --ribbon_starting_level=999 --secondary_cache_fault_one_in=32 --secondary_cache_uri=compressed_secondary_cache://capacity=8388608 --set_options_one_in=10000 --snapshot_hold_ops=100000 --sst_file_manager_bytes_per_sec=0 --sst_file_manager_bytes_per_truncate=0 --subcompactions=3 --sync_fault_injection=1 --target_file_size_base=2097 --target_file_size_multiplier=2 --test_batches_snapshots=1 --top_level_index_pinning=1 --use_full_merge_v1=1 --use_merge=1 --value_size_mult=32 --verify_checksum=1 --verify_checksum_one_in=1000000 --verify_db_one_in=100000 --verify_sst_unique_id_in_manifest=1 --wal_bytes_per_sync=524288 --write_buffer_size=4194 --writepercent=35
```

```
stderr:
WARNING: prefix_size is non-zero but memtablerep != prefix_hash
db_stress: utilities/fault_injection_fs.cc:748: virtual rocksdb::IOStatus rocksdb::FaultInjectionTestFS::RenameFile(const std::string &, const std::string &, const rocksdb::IOOptions &, rocksdb::IODebugContext *): Assertion `tlist.find(tdn.second) == tlist.end()' failed.`
```

**Summary:**
The PR ensured the non-test path pass down a non-null dir containing CURRENT (which is by current RocksDB assumption just db_dir) by doing the following:
- Renamed `directory_to_fsync` as `dir_contains_current_file` in `SetCurrentFile()` to tighten the association between this directory and CURRENT file
- Changed `SetCurrentFile()` API to require `dir_contains_current_file` being passed-in, instead of making it by default nullptr. 
    -  Because `SetCurrentFile()`'s `dir_contains_current_file` is passed down from `VersionSet::LogAndApply()` then `VersionSet::ProcessManifestWrites()` (i.e, think about this as a chain of 3 functions related to MANIFEST update), these 2 functions also got refactored to require `dir_contains_current_file`
- Updated the non-test-path callers of these 3 functions to obtain and pass in non-nullptr `dir_contains_current_file`, which by current assumption of RocksDB, is the `FSDirectory* db_dir`.
    - `db_impl` path will obtain `DBImpl::directories_.getDbDir()` while others with no access to such `directories_` are obtained on the fly by creating such object `FileSystem::NewDirectory(..)` and manage it by unique pointers to ensure short life time. 

**Test:**
- `make check`
- Passed the repro db_stress command
- For future improvement, since we currently don't assert dir containing CURRENT to be non-nullptr due to https://github.com/facebook/rocksdb/pull/10573#pullrequestreview-1087698899, there is still chances that future developers mistakenly pass down nullptr dir containing CURRENT thus resulting skipped sync dir and cause the bug again. Therefore a smarter test (e.g, such as quoted from @ajkr  "(make) unsynced data loss to be dropping files corresponding to unsynced directory entries") is still needed.

### CODE REVIEW

File: db/version_set.cc [hunk 1]
```diff
@@ -4892,6 +4892,13 @@ Status VersionSet::ProcessManifestWrites(
       assert(manifest_io_status.ok());
     }
     if (s.ok() && new_descriptor_log) {
+      Directories directories_;
+      if (db_directory == nullptr) {
```
Comment (ajkr) [inline db/version_set.cc:4896]: Why would db_directory be nullptr in the depths of ProcessManifestWrites()? Feels like callers might not be passing db_directory properly.

File: db/db_impl/db_impl.cc [hunk 2]
```diff
@@ -1494,8 +1494,8 @@ Status DBImpl::SyncWAL() {
 Status DBImpl::ApplyWALToManifest(VersionEdit* synced_wals) {
   // not empty, write to MANIFEST.
   mutex_.AssertHeld();
-  Status status =
-      versions_->LogAndApplyToDefaultColumnFamily(synced_wals, &mutex_);
+  Status status = versions_->LogAndApplyToDefaultColumnFamily(
+      synced_wals, &mutex_, directories_.GetDbDir());
```
Comment (ajkr) [inline db/db_impl/db_impl.cc:1498]: Makes sense.

File: db/db_impl/db_impl_open.cc [hunk 3]
```diff
@@ -340,6 +340,10 @@ Status DBImpl::NewDB(std::vector<std::string>* new_filenames) {
   }
   if (s.ok()) {
     // Make "CURRENT" file that points to the new manifest file.
+    s = directories_.SetDirectories(fs_.get(), dbname_,
+                                    immutable_db_options_.wal_dir,
+                                    immutable_db_options_.db_paths);
+    assert(s.ok());
```
Comment (ajkr) [inline db/db_impl/db_impl_open.cc:346]: I checked `Recover()` does this before it calls `NewDB()`. Why is this needed?

**Comment by hx235** [comment]
Update:
- Still a draft with a lot more refectory adapting to the LogAndApply->ProcessManifestWrites->SetCurrentFile API changes.

File: db/version_set.cc (hunk 1 above)
Comment (hx235) [reply db/version_set.cc]: Fixed by requiring such nullptr in upper callers.

File: db/db_impl/db_impl_open.cc (hunk 3 above)
Comment (hx235) [reply db/db_impl/db_impl_open.cc:346]: Yeah - it's a bit messy here and just for hack. See PR summary for the problem related to it.

Comment (ajkr) [reply db/db_impl/db_impl_open.cc:346]: Do you mean this part: "it's because they are either just unit tests or internal logic where directories_ has not yet set properly with non-null db dir"? I don't know what is the code path where directories_ isn't set properly. DB logic looks ok thanks to https://github.com/facebook/rocksdb/blob/d17be55aab80b856f96f4af89f8d18fef96646b4/db/db_impl/db_impl_open.cc#L413-L415. Only other caller looks like repair tool, which seems able to set up directories_ if not already

File: db/version_set.cc [hunk 4]
```diff
@@ -5642,9 +5643,14 @@ Status VersionSet::ReduceNumberOfLevels(const std::string& dbname,
...
-      versions.GetColumnFamilySet()->GetDefault(),
-      mutable_cf_options, &ve, &dummy_mutex, nullptr, true);
+  std::unique_ptr<FSDirectory> db_dir;
+  assert((versions.env_)
+             ->GetFileSystem()
+             ->NewDirectory(dbname, IOOptions(), &db_dir, nullptr)
+             .ok());
+  return versions.LogAndApply(versions.GetColumnFamilySet()->GetDefault(),
+                              mutable_cf_options, &ve, &dummy_mutex,
+                              db_dir.get(), true);
```
Comment (ajkr) [inline db/version_set.cc:5653]: We can't do this (extraneous NewDirectory or critical logic in an assert)

File: tools/ldb_cmd.cc [hunk 5]
```diff
@@ -4196,7 +4196,11 @@ void UnsafeRemoveSstFileCommand::DoCommand() {
     VersionEdit edit;
     edit.SetColumnFamily(cfd->GetID());
     edit.DeleteFile(level, sst_file_number_);
-    s = w.LogAndApply(cfd, &edit);
+    std::unique_ptr<FSDirectory> db_dir_;
+    assert(options_.env->GetFileSystem()
+               ->NewDirectory(db_path_, IOOptions(), &db_dir_, nullptr)
+               .ok());
+    s = w.LogAndApply(cfd, &edit, db_dir_.get());
```
Comment (ajkr) [inline tools/ldb_cmd.cc:4203]: Fine with me except the `assert()` and `db_dir_` is named like an instance variable.

**Review by ajkr — CHANGES_REQUESTED** [review]
I wouldn't mind if the scope of this PR narrowed to eliminating the default value (nullptr) for CURRENT directory argument and making sure that non-test code always provides something reasonable for that argument (e.g., `directories_.GetDbDir()`). Disallowing nullptr in `SetCurrentFile()` seems like extra work (lower value, IMO) but fine if you want to do it.

File: db/db_impl/db_impl_open.cc (hunk 3 above)
Comment (hx235) [reply db/db_impl/db_impl_open.cc:346]: > Only other caller looks like repair tool, which seems able to set up directories_ if not already

This was hack that I used on my first draft to pass the all non-nullptr assertion for all tests. I can't remember exactly which UT now as we are removing that assertion and re-reading the code path seems fine with me without this hack.

File: tools/ldb_cmd.cc (hunk 5 above)
Comment (hx235) [reply tools/ldb_cmd.cc:4203]: Fixed

File: db/repair.cc [hunk 6]
```diff
@@ -656,9 +660,13 @@ class Repairer {
...
-      Status status = vset_.LogAndApply(
-          cfd, *cfd->GetLatestMutableCFOptions(), &edit, &mutex_,
-          nullptr /* db_directory */, false /* new_descriptor_log */);
+      std::unique_ptr<FSDirectory> db_dir;
+      assert(env_->GetFileSystem()
+                 ->NewDirectory(dbname_, IOOptions(), &db_dir, nullptr)
+                 .ok());
+      Status status = vset_.LogAndApply(cfd, *cfd->GetLatestMutableCFOptions(),
+                                        &edit, &mutex_, db_dir.get(),
+                                        false /* new_descriptor_log */);
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/repair.cc:669]: Fixed

File: db/repair.cc [hunk 7]
```diff
@@ -162,9 +162,13 @@ class Repairer {
...
-    Status status = vset_.LogAndApply(cfd, mut_cf_opts, &edit, &mutex_,
-                                      nullptr /* db_directory */,
-                                      false /* new_descriptor_log */, cf_opts);
+    std::unique_ptr<FSDirectory> db_dir;
+    assert(env_->GetFileSystem()
+               ->NewDirectory(dbname_, IOOptions(), &db_dir, nullptr)
+               .ok());
+    Status status =
+        vset_.LogAndApply(cfd, mut_cf_opts, &edit, &mutex_, db_dir.get(),
+                          false /* new_descriptor_log */, cf_opts);
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/repair.cc:171]: Fixed

File: db/experimental.cc [hunk 8]
```diff
@@ -122,7 +122,9 @@ Status UpdateManifestForFilesState(
     }
 
     if (s.ok() && edit.NumEntries() > 0) {
-      s = w.LogAndApply(cfd, &edit);
+      std::unique_ptr<FSDirectory> db_dir_;
+      assert(fs->NewDirectory(db_name, IOOptions(), &db_dir_, nullptr).ok());
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/experimental.cc:126]: Fixed

File: db/db_impl/db_impl.cc [hunk 9]
```diff
@@ -3148,8 +3149,10 @@ Status DBImpl::DropColumnFamilyImpl(ColumnFamilyHandle* column_family) {
       // we drop column family from a single write thread
       WriteThread::Writer w;
       write_thread_.EnterUnbatched(&w, &mutex_);
+      std::unique_ptr<FSDirectory> db_dir_;
+      assert(fs_->NewDirectory(dbname_, IOOptions(), &db_dir_, nullptr).ok());
       s = versions_->LogAndApply(cfd, *cfd->GetLatestMutableCFOptions(), &edit,
-                                 &mutex_);
+                                 &mutex_, db_dir_.get());
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/db_impl/db_impl.cc:3155]: Fixed

File: db/db_impl/db_impl.cc [hunk 10]
```diff
@@ -3046,9 +3046,10 @@ Status DBImpl::CreateColumnFamilyImpl(const ColumnFamilyOptions& cf_options,
       write_thread_.EnterUnbatched(&w, &mutex_);
       // LogAndApply will both write the creation in MANIFEST and create
       // ColumnFamilyData object
+      std::unique_ptr<FSDirectory> db_dir_;
+      assert(fs_->NewDirectory(dbname_, IOOptions(), &db_dir_, nullptr).ok());
       s = versions_->LogAndApply(nullptr, MutableCFOptions(cf_options), &edit,
-                                 &mutex_, directories_.GetDbDir(), false,
-                                 &cf_options);
+                                 &mutex_, db_dir_.get(), false, &cf_options);
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/db_impl/db_impl.cc:3052]: Fixed

File: db/db_impl/db_impl_open.cc (hunk 3 above)
Comment (hx235) [reply db/db_impl/db_impl_open.cc:346]: Fixed

**Comment by hx235** [comment]
> I wouldn't mind if the scope of this PR narrowed to eliminating the default value (nullptr) for CURRENT directory argument and making sure that non-test code always provides something reasonable for that argument (e.g., `directories_.GetDbDir()`). Disallowing nullptr in `SetCurrentFile()` seems like extra work (lower value, IMO) but fine if you want to do it.

Narrowed the scope of the PR for quicker turn-around.

Also renamed the current_file_dir to be dir_contains_current_file to be more self-explanatory.

**Review by ajkr — APPROVED** [review]
LGTM, thanks!!
